### PR TITLE
Dataset api restructuring

### LINKED ISSUE #269: rethink Dataset vs MagDataset names and usage in DatasetApi
When speaking or thinking about the dataset api, I find the terms `Dataset` and `MagDataset` quite confusing. My current thoughts:
- Just rename `MagDataset` to `Mag`, but this conflicts with the existing `Mag`, and does not carry the notion of being write/readable. `add_mag` and other layer-methods have the same problem.
- Call it `LayerMag`, which might lead to confusion with the `Layer` again, but seems a little better to me.
- Rethink how to interact with the API first, and then think about good names. E.g. as a user, which is the "standard" way to open a dataset and get or write some data:
  * Should I always use the `Dataset` and its methods, and just get a view from it?
  * Should I always work with a view-object and just switch from there?
  * Or rather create a dataset, a layer, a mag, a view, and then work with it?
  * …
  Having multiple options might be fine, but then we should advocate a simple default clearly.

### LINKED ISSUE #327: use standard lowercase modulenames
According to [PEP8](https://www.python.org/dev/peps/pep-0008/#package-and-module-names) module-names (aka .py-filenames) should be snakecased (all lowercase using `_` as a delimiter). This also is a standard across most python packages.

### LINKED ISSUE #328: Shorthands for read/write in Dataset API
From the #302 PR I learned about the `Dataset.get_view` shorthand. That made me wonder if a shorthand for the frequently used `read` and `write` methods would also make sense.

```python
ds = WKDataset.create(...)
ds.write(layer='color', layer_category='color', dtype=np.float32, mag='2-2-1', offset=(1024, 1024, 1024), data=data)
data = ds.read(layer='color', mag='2-2-1', offset=(1024, 1024, 1024), shape=(32, 32, 32))
```

Any thoughts? @rschwanhold @philippotto @jstriebel

### PR DESCRIPTION
**Pull request by rschwanhold**
### Description:
- Due to the lack of usage, we decided to remove the support for tiffs in the dataset API. This removes a lot of code.
- There were multiple smaller things that are now fixed (e.g. lower case module names)
- The `WKMagDataset` previously delegated many operations directly to the `View`. This is now refactored by inheriting from `View` directly.

### Issues:
- fixes #327 
- fixes #328 
- fixes #269

### Todos:
Make sure to delete unnecessary points or to check all before merging:
 - [x] Updated Changelog
 - [x] Fix tests after we have agreed on the behavior of `Views`

### CODE REVIEW

**Comment by rschwanhold** [comment]
This PR refactors the dataset API and introduces some breaking changes. While we are at it, there is the issue #269 about the naming. I think now would be a good time to discuss it. So far I left the name like they used to be (e.g. `WKDataset`). However the prefix `WK` was only necessary to distinguish between `WK` and `Tiff`, which is no longer the case.

A large part of this refactoring deals with the `WKMagDataset`. The `WKMagDataset` inherits from the `View`. `Views` are now always bounded while the `WKMagDataset` is not (this means that read/write operations of a `WKMagDataset` may exceed the specified bounding box).

This raises the questing how strict `WKMagDataset.get_view()` should be:
- Option 1: `WKMagDataset.get_view()` may exceed the specified bounding box (just like read or write). The donwside is that a `View` no longer has a reference to the properties. Therefore, the bounding boxes cannot be updated (if necessary).
- Option 2: `WKMagDataset.get_view()` may not exceed the specified bounding box (just like `View.get_view()`). This would be safe because the user cannot get a view that is outside the bounding box (-> prevents that write-operations can cause an inconsistent state in the properties)
- Option 3: `WKMagDataset.get_view()` may only exceed the specified bounding box if `read_only=True`. This prevents that write-operations can cause an inconsistent state in the properties. Still being able to get larger views to read data can be very helpful because `wkw.Dataset` automatically pads the data with `0`. We currently make use of this in the downsampling-code, which leads to nice code.

Note: `View.get_view()` only allows sub-views because a `View` is bounded.

What are your thought on the naming and the behavior of the Views?

cc: @philippotto @normanrz @jstriebel

**Comment by jstriebel** [comment]
Notes from a call with @rschwanhold just now: Option 3 sounds nice and fits our use-cases best, without doing unexpected things. For the naming `Dataset`, `Layer`, `MagView` and `View` seem fitting. `MagView` has at least the same capabilities as a view, so this seems better than the current `MagDataset`, since it's quit different from `Dataset`.

**Comment by rschwanhold** [comment]
Regarding `get_view()`, @jstriebel and I agreed that option 3 seems to be the best.

Regarding the naming:
We could rename `WKDataset` simply to `Dataset` and `WKMagDataset` to `MagView`. This would prevent that both names include the term "Dataset". The name `MagView` would fit very well because this PR implements the old `MagDataset` by inheriting from `View`. 
The structure of the dataset API would then be: `Dataset` -> `Layer` -> `MagView`

I am still open for other ideas.

File: wkcuber/convert_nifti.py
Comment (philippotto) [inline wkcuber/convert_nifti.py:75]: Are we sure that this is not used? @leowe You are probably the one who can decide best whether this parameter can be removed.

**Review by philippotto — CHANGES_REQUESTED** [review]
Whew, quite a bit of changes! Great work overall :+1: Looks very good. I left a few comments, but nothing major. I hope that the decision to drop the tiff support doesn't bite us in the future..

File: wkcuber/api/mag_view.py
Comment (rschwanhold) [reply wkcuber/api/mag_view.py]: Yes, you are right. A `MagView` always deals with the whole magnification, whereas `View`s only look at sub-regions.
I have updated the documentation

**Comment by rschwanhold** [comment]
@normanrz can you also take a second look at the documentation to make sure that this meets your expectations?

**Comment by normanrz** [comment]
I get an error when trying to build the docs:
```
(wk) ~/scalableminds/tools/webknossos-cuber dataset-api-restructuring → docs/api.sh --persist
/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py:130: RuntimeWarning: Error loading wkcuber:
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 219, in load_module
    return importlib.import_module(module)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "./wkcuber/__init__.py", line 9, in <module>
    from .api.dataset import Dataset
  File "./wkcuber/api/dataset.py", line 13, in <module>
    from wkcuber.api.properties.layer_properties import (
ModuleNotFoundError: No module named 'wkcuber.api.properties'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 128, in walk_packages2
    module = load_module(mod.name)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 221, in load_module
    raise RuntimeError(f"Error importing {module}") from e
RuntimeError: Error importing wkcuber

  warnings.warn(
/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py:405: RuntimeWarning: Error importing wkcuber:
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 219, in load_module
    return importlib.import_module(module)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "./wkcuber/__init__.py", line 9, in <module>
    from .api.dataset import Dataset
  File "./wkcuber/api/dataset.py", line 13, in <module>
    from wkcuber.api.properties.layer_properties import (
ModuleNotFoundError: No module named 'wkcuber.api.properties'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py", line 403, in pdoc
    m = extract.load_module(mod)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 221, in load_module
    raise RuntimeError(f"Error importing {module}") from e
RuntimeError: Error importing wkcuber

  warnings.warn(
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/wk/bin/pdoc", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__main__.py", line 153, in cli
    pdoc.pdoc(
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py", line 420, in pdoc
    (output_directory / "index.html").write_bytes(index.encode())
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1244, in write_bytes
    with self.open(mode='wb') as f:
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1221, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1077, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: 'docs/api/index.html'
```

**Comment by rschwanhold** [comment]
> I get an error when trying to build the docs:
> 
> ```
> (wk) ~/scalableminds/tools/webknossos-cuber dataset-api-restructuring → docs/api.sh --persist
> /usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py:130: RuntimeWarning: Error loading wkcuber:
> Traceback (most recent call last):
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 219, in load_module
>     return importlib.import_module(module)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/importlib/__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "./wkcuber/__init__.py", line 9, in <module>
>     from .api.dataset import Dataset
>   File "./wkcuber/api/dataset.py", line 13, in <module>
>     from wkcuber.api.properties.layer_properties import (
> ModuleNotFoundError: No module named 'wkcuber.api.properties'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 128, in walk_packages2
>     module = load_module(mod.name)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/contextlib.py", line 75, in inner
>     return func(*args, **kwds)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 221, in load_module
>     raise RuntimeError(f"Error importing {module}") from e
> RuntimeError: Error importing wkcuber
> 
>   warnings.warn(
> /usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py:405: RuntimeWarning: Error importing wkcuber:
> Traceback (most recent call last):
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 219, in load_module
>     return importlib.import_module(module)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/importlib/__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "./wkcuber/__init__.py", line 9, in <module>
>     from .api.dataset import Dataset
>   File "./wkcuber/api/dataset.py", line 13, in <module>
>     from wkcuber.api.properties.layer_properties import (
> ModuleNotFoundError: No module named 'wkcuber.api.properties'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py", line 403, in pdoc
>     m = extract.load_module(mod)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/contextlib.py", line 75, in inner
>     return func(*args, **kwds)
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/extract.py", line 221, in load_module
>     raise RuntimeError(f"Error importing {module}") from e
> RuntimeError: Error importing wkcuber
> 
>   warnings.warn(
> Traceback (most recent call last):
>   File "/usr/local/anaconda3/envs/wk/bin/pdoc", line 8, in <module>
>     sys.exit(cli())
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__main__.py", line 153, in cli
>     pdoc.pdoc(
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/site-packages/pdoc/__init__.py", line 420, in pdoc
>     (output_directory / "index.html").write_bytes(index.encode())
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1244, in write_bytes
>     with self.open(mode='wb') as f:
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1221, in open
>     return io.open(self, mode, buffering, encoding, errors, newline,
>   File "/usr/local/anaconda3/envs/wk/lib/python3.8/pathlib.py", line 1077, in _opener
>     return self._accessor.open(self, flags, mode)
> FileNotFoundError: [Errno 2] No such file or directory: 'docs/api/index.html'
> ```

I don't quite understand your command line (`~/scalableminds/tools/webknossos-cuber dataset-api-restructuring → docs/api.sh --persist`). Is this some plugin that shows which branch you are on?

If I am in the root dire of the repo, executing `./docs/api.sh` (or `./docs/api.sh --persist`) works. 

@philippotto do you have the same problem?

**Comment by normanrz** [comment]
>  I don't quite understand your command line (~/scalableminds/tools/webknossos-cuber dataset-api-restructuring → docs/api.sh --persist). Is this some plugin that shows which branch you are on?

Yeah. It shows the branch. I was in the root of the repo. I did `poetry install` and ran `./docs/api.sh --persist`

**Comment by philippotto** [comment]
> If I am in the root dire of the repo, executing `./docs/api.sh` (or `./docs/api.sh --persist`) works.
> 
> @philippotto do you have the same problem?

Works for me, too. I tested it in docker. I get some warnings, though:

```
root@4d8c117bef85:/app# ./docs/api.sh --persist
/usr/local/lib/python3.7/site-packages/pdoc/doc_types.py:121: UserWarning: Error parsing type annotation for wkcuber.api.layer.Layer.__init__. Import of Dataset failed: name 'Dataset' is not defined
  f"Error parsing type annotation for {fullname}. Import of {mod} failed: {err}"
/usr/local/lib/python3.7/site-packages/pdoc/doc_types.py:121: UserWarning: Error parsing type annotation for wkcuber.api.mag_view.MagView.__init__. Import of Layer failed: name 'Layer' is not defined
  f"Error parsing type annotation for {fullname}. Import of {mod} failed: {err}"
```

The docs are generated successfully, though.

**Review by philippotto — APPROVED** [review]
looks great to me :)

**Comment by rschwanhold** [comment]
> > If I am in the root dire of the repo, executing `./docs/api.sh` (or `./docs/api.sh --persist`) works.
> > @philippotto do you have the same problem?
> 
> Works for me, too. I tested it in docker. I get some warnings, though:
> 
> ```
> root@4d8c117bef85:/app# ./docs/api.sh --persist
> /usr/local/lib/python3.7/site-packages/pdoc/doc_types.py:121: UserWarning: Error parsing type annotation for wkcuber.api.layer.Layer.__init__. Import of Dataset failed: name 'Dataset' is not defined
>   f"Error parsing type annotation for {fullname}. Import of {mod} failed: {err}"
> /usr/local/lib/python3.7/site-packages/pdoc/doc_types.py:121: UserWarning: Error parsing type annotation for wkcuber.api.mag_view.MagView.__init__. Import of Layer failed: name 'Layer' is not defined
>   f"Error parsing type annotation for {fullname}. Import of {mod} failed: {err}"
> ```
> 
> The docs are generated successfully, though.

Yes, I get the warnings too. The reason for this is that the import of e.g. `Dataset` in `Layer` only happens if `TYPE_CHECKING` is `True`. This flag is set by mypy. Otherwise there would be circular dependencies. Apparently, `pdoc` does not set this flag. As a result, we get the warnings.

**Comment by philippotto** [comment]
reminder that https://github.com/scalableminds/webknossos-cuber/pull/331#discussion_r657150036 should be clarified before merging.

File: wkcuber/convert_nifti.py
Comment (leowe) [reply wkcuber/convert_nifti.py:75]: I think this can be removed. I used it only for debug purposes and as far as I know we currently have no direct use case for it in a project context.

**Comment by rschwanhold** [comment]
@normanrz do you still have problems with `pdoc`? I am using conda with python `3.7`. My `pdoc` version is `7.0.1`.

**Comment by normanrz** [comment]
I still have the issue. I run python 3.8. But, I don't think this needs to block the merge.

**Pull request merged**
